File: Day 29 - Number Theory (part 2)/prime sum.cpp
#include <bits/stdc++.h>
using namespace std;

#define N 1000000 + 1

void prime_sieve(vector<int> &sieve)
{

  // mark 1 and 0 as non-prime
  sieve[0] = sieve[1] = 0;
  ;

  for (long long p = 2; p < N; p++)
  {
    if (sieve[p] == 1)
    {
      for (long long j = p * p; j < N; j += p)
      {
        // marking j as non-prime
        sieve[j] = 0;
      }
    }
  }
}

int main()
{
  int n;
  cin >> n;
  vector<int> sieve(N, 1);
  prime_sieve(sieve);
  int ans = 0;
  for (int i = 3; i <= n - 2; i++)
  {
    if (sieve[i] == 1 && sieve[i + 2] == 1)
    {
      ans++;
    }
  }
  cout << ans;
  return 0;
}

File: Day 29 - Number Theory (part 2)/prime sum.java
import java.util.Scanner;

public class Main {
  static final int N = 1000000 + 1;

  static void primeSieve(int[] sieve) {
    sieve[0] = sieve[1] = 0;

    for (int p = 2; p < N; p++) {
      if (sieve[p] == 1) {
        for (long j = (long) p * p; j < N; j += p) {
          sieve[(int) j] = 0;
        }
      }
    }
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    int[] sieve = new int[N];
    for (int i = 0; i < N; i++) {
      sieve[i] = 1;
    }
    primeSieve(sieve);
    int ans = 0;
    for (int i = 3; i <= n - 2; i++) {
      if (sieve[i] == 1 && sieve[i + 2] == 1) {
        ans++;
      }
    }
    System.out.println(ans);
    scanner.close();
  }
}
